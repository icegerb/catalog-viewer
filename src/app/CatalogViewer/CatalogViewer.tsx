'use client';

import React, { useEffect, useState } from 'react';
import { IImage } from '@/types/image';
import { FaArrowLeft, FaArrowRight } from 'react-icons/fa';
import Viewer from './Viewer';
import Thumbs from './Thumbs';
import IconButton from '../components/IconButton';

const images: IImage[] = [
  {
    name: 'candle holder 1',
    url: '/images/1.jpeg',
  },
  {
    name: 'candle holder 2',
    url: '/images/2.jpeg',
  },
  {
    name: 'candle holder 3',
    url: '/images/3.jpeg',
  },
  {
    name: 'candle holder 4',
    url: '/images/4.jpeg',
  },
];

const CatalogViewer = () => {
  const [isChecked, setIsChecked] = useState<boolean>(false);
  const [selectedImage, setSelectedImage] = useState<IImage>(images[0]);

  const slideNavigation = (direction: 'prev' | 'next') => {
    if (!images || images.length === 0) return;

    setSelectedImage((prev) => {
      const currentIndex = images.indexOf(prev);

      const newIndex =
        direction === 'next'
          ? (currentIndex + 1) % images.length
          : (currentIndex - 1 + images.length) % images.length;
      return images[newIndex];
    });
  };

  useEffect(() => {
    if (isChecked) {
      const interval = setInterval(() => slideNavigation('next'), 3000);
      return () => clearInterval(interval);
    }
  }, [isChecked]);

  return (
    <div className='shadow-md pb-4'>
      <h2 className='text-l bg-slate-600 text-green-500 p-2 text-center font-bold'>
        Catalog Viewer
      </h2>
      <div className='shadow-md mx-8 mt-12 mb-4 flex flex-col gap-5 border pt-5 pb-2 px-2'>
        <Viewer image={selectedImage} />
        <div className='flex justify-between items-center'>
          <IconButton
            icon={<FaArrowLeft />}
            onClick={() => slideNavigation('prev')}
            dataTestId={'prev-slide-btn'}
          />
          <Thumbs images={images} setSelectedImage={setSelectedImage} />
          <IconButton
            icon={<FaArrowRight />}
            onClick={() => slideNavigation('next')}
            dataTestId={'next-slide-btn'}
          />
        </div>
      </div>

      <form className='flex justify-center gap-2'>
        <input
          type='checkbox'
          id='slideShow'
          name='slideShow'
          checked={isChecked}
          onChange={(e) => setIsChecked(e.target.checked)}
          data-testid='slide-show-checkbox'
        />
        <label htmlFor='slideShow'>Start Slide Show</label>
      </form>
    </div>
  );
};

export default CatalogViewer;
