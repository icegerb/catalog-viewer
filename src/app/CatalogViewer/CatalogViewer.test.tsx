import React from 'react';
import { IImage } from '@/types/image';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import CatalogViewer from './CatalogViewer';

const mockImages: IImage[] = [
  { name: 'candle holder 1', url: '/images/1.jpeg' },
  { name: 'candle holder 2', url: '/images/2.jpeg' },
  { name: 'candle holder 3', url: '/images/3.jpeg' },
  { name: 'candle holder 4', url: '/images/4.jpeg' },
];

describe('CatalogViewer component', () => {
  beforeEach(() => {
    render(<CatalogViewer />);
  });

  it('renders all elements in CatalogViewer', () => {
    expect(
      screen.getByAltText(`image-${mockImages[0].name}`)
    ).toBeInTheDocument();
    expect(screen.getByTestId('next-slide-btn')).toBeInTheDocument();
    expect(screen.getByTestId('prev-slide-btn')).toBeInTheDocument();
    expect(screen.getByTestId('slide-show-checkbox')).toBeInTheDocument();
  });

  it('should navigate images when next and prev buttons are selected', () => {
    const prevBtn = screen.getByTestId('prev-slide-btn');
    const nextBtn = screen.getByTestId('next-slide-btn');

    fireEvent.click(nextBtn);
    expect(screen.getByAltText('candle holder 2')).toBeInTheDocument();

    fireEvent.click(prevBtn);
    expect(screen.getByAltText('candle holder 1')).toBeInTheDocument();
  });

  it('should toggle the image slide show when checkbox is checked', async () => {
    const checkbox = screen.getByTestId('slide-show-checkbox');
    expect(checkbox).not.toBeChecked();
    fireEvent.click(checkbox);
    expect(checkbox).toBeChecked();
    await waitFor(
      () => expect(screen.getByAltText('candle holder 2')).toBeInTheDocument(),
      { timeout: 3500 }
    );
  });
});
