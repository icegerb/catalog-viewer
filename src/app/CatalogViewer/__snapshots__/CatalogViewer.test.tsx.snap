// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogViewer component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="shadow-md pb-4"
  >
    <h2
      class="text-l bg-slate-600 text-green-500 p-2 text-center font-bold"
    >
      Catalog Viewer
    </h2>
    <div
      class="shadow-md mx-8 mt-12 mb-4 flex flex-col gap-5 border pt-5 pb-2 px-2"
    >
      <div
        class="px-8"
        data-testid="catalog-view"
      >
        <div
          class="w-[300px] h-[300px] overflow-hidden"
        >
          <img
            alt="image-candle holder 1"
            class="w-full h-full object-cover object-center"
            data-nimg="1"
            decoding="async"
            height="500"
            src="/_next/image?url=%2Fimages%2F1.jpeg&w=1080&q=75"
            srcset="/_next/image?url=%2Fimages%2F1.jpeg&w=640&q=75 1x, /_next/image?url=%2Fimages%2F1.jpeg&w=1080&q=75 2x"
            style="color: transparent;"
            width="500"
          />
        </div>
      </div>
      <div
        class="flex justify-between items-center"
      >
        <button
          class="flex rounded-full border-2 border-slate-300 border-solid p-1 text-green-700 h-7 w-7 justify-center items-center hover:bg-green-700 hover:text-white hover:border-green-700 >"
          data-testid="prev-slide-btn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
            />
          </svg>
        </button>
        <div
          class="relative"
        >
          <button
            class="absolute hover:border-solid hover:border-gray-500 hover:border-2 w-[80px] h-[75px] top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"
            data-testid="thumb-button-0"
          />
          <div
            class="w-16 h-16"
          >
            <img
              alt="candle holder 1"
              class="w-full h-full object-cover object-center"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2F1.jpeg&w=256&q=75"
              srcset="/_next/image?url=%2Fimages%2F1.jpeg&w=128&q=75 1x, /_next/image?url=%2Fimages%2F1.jpeg&w=256&q=75 2x"
              style="color: transparent;"
              width="100"
            />
          </div>
        </div>
        <div
          class="relative"
        >
          <button
            class="absolute hover:border-solid hover:border-gray-500 hover:border-2 w-[80px] h-[75px] top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"
            data-testid="thumb-button-1"
          />
          <div
            class="w-16 h-16"
          >
            <img
              alt="candle holder 2"
              class="w-full h-full object-cover object-center"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2F2.jpeg&w=256&q=75"
              srcset="/_next/image?url=%2Fimages%2F2.jpeg&w=128&q=75 1x, /_next/image?url=%2Fimages%2F2.jpeg&w=256&q=75 2x"
              style="color: transparent;"
              width="100"
            />
          </div>
        </div>
        <div
          class="relative"
        >
          <button
            class="absolute hover:border-solid hover:border-gray-500 hover:border-2 w-[80px] h-[75px] top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"
            data-testid="thumb-button-2"
          />
          <div
            class="w-16 h-16"
          >
            <img
              alt="candle holder 3"
              class="w-full h-full object-cover object-center"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2F3.jpeg&w=256&q=75"
              srcset="/_next/image?url=%2Fimages%2F3.jpeg&w=128&q=75 1x, /_next/image?url=%2Fimages%2F3.jpeg&w=256&q=75 2x"
              style="color: transparent;"
              width="100"
            />
          </div>
        </div>
        <div
          class="relative"
        >
          <button
            class="absolute hover:border-solid hover:border-gray-500 hover:border-2 w-[80px] h-[75px] top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"
            data-testid="thumb-button-3"
          />
          <div
            class="w-16 h-16"
          >
            <img
              alt="candle holder 4"
              class="w-full h-full object-cover object-center"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2F4.jpeg&w=256&q=75"
              srcset="/_next/image?url=%2Fimages%2F4.jpeg&w=128&q=75 1x, /_next/image?url=%2Fimages%2F4.jpeg&w=256&q=75 2x"
              style="color: transparent;"
              width="100"
            />
          </div>
        </div>
        <button
          class="flex rounded-full border-2 border-slate-300 border-solid p-1 text-green-700 h-7 w-7 justify-center items-center hover:bg-green-700 hover:text-white hover:border-green-700 >"
          data-testid="next-slide-btn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
            />
          </svg>
        </button>
      </div>
    </div>
    <form
      class="flex justify-center gap-2"
    >
      <input
        data-testid="slide-show-checkbox"
        id="slideShow"
        name="slideShow"
        type="checkbox"
      />
      <label
        for="slideShow"
      >
        Start Slide Show
      </label>
    </form>
  </div>
</DocumentFragment>
`;
